File: cook/loop.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"text/template"
	"time"
)

type Config struct {
	WorkPrompt    string
	ReviewPrompt  string
	GatePrompt    string
	MaxIterations int
	Model         string
	ProjectRoot   string
}

func agentLoop(sandbox *Sandbox, config Config, cookMD string) error {
	tmpl, err := template.New("cook").Parse(cookMD)
	if err != nil {
		return fmt.Errorf("parsing COOK.md: %w", err)
	}

	logFile, err := createSessionLog(config.ProjectRoot)
	if err != nil {
		return err
	}
	logOK("Session log: %s", logFile)
	logStep("Follow along: tail -f %s", logFile)

	lastMessage := ""

	for i := 1; i <= config.MaxIterations; i++ {
		steps := []struct {
			name   string
			prompt string
		}{
			{"work", config.WorkPrompt},
			{"review", config.ReviewPrompt},
			{"gate", config.GatePrompt},
		}

		for _, step := range steps {
			logPhase("Iteration %d/%d — %s", i, config.MaxIterations, step.name)

			prompt, err := executeTemplate(tmpl, LoopContext{
				Step:          step.name,
				Prompt:        step.prompt,
				LastMessage:   lastMessage,
				Iteration:     i,
				MaxIterations: config.MaxIterations,
				LogFile:       logFile,
			})
			if err != nil {
				return err
			}

			output, err := sandbox.runClaude(config.Model, prompt)
			if err != nil {
				return fmt.Errorf("%s step failed (iteration %d): %w", step.name, i, err)
			}

			lastMessage = output
			if err := appendToLog(logFile, step.name, i, output); err != nil {
				logWarn("Failed to write session log: %v", err)
			}
		}

		// Check gate verdict
		verdict := parseGateVerdict(lastMessage)
		if verdict == "DONE" {
			logOK("Gate: DONE — loop complete")
			return nil
		}
		if i < config.MaxIterations {
			logWarn("Gate: ITERATE — continuing to iteration %d", i+1)
		} else {
			logWarn("Gate: max iterations (%d) reached — stopping", config.MaxIterations)
		}
	}
	return nil
}

func executeTemplate(tmpl *template.Template, ctx LoopContext) (string, error) {
	var buf strings.Builder
	if err := tmpl.Execute(&buf, ctx); err != nil {
		return "", fmt.Errorf("rendering COOK.md: %w", err)
	}
	return buf.String(), nil
}

// createSessionLog returns a host filesystem path that is also valid inside
// the container because the project root is bind-mounted at the same path.
func createSessionLog(projectRoot string) (string, error) {
	logDir := filepath.Join(projectRoot, ".cook", "logs")
	if err := os.MkdirAll(logDir, 0755); err != nil {
		return "", fmt.Errorf("creating log directory: %w", err)
	}
	timestamp := time.Now().Format("2006-01-02-150405")
	logPath := filepath.Join(logDir, timestamp+".md")
	f, err := os.Create(logPath)
	if err != nil {
		return "", fmt.Errorf("creating log file: %w", err)
	}
	f.Close()
	return logPath, nil
}

func appendToLog(logFile, step string, iteration int, output string) error {
	f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = fmt.Fprintf(f, "## [%s %d] %s\n\n%s\n\n---\n\n",
		step, iteration, time.Now().Format("2006-01-02 15:04:05"), output)
	return err
}

var doneKeywords = []string{"DONE", "PASS", "COMPLETE", "APPROVE", "ACCEPT"}
var iterateKeywords = []string{"ITERATE", "REVISE", "RETRY"}

func parseGateVerdict(output string) string {
	scanner := bufio.NewScanner(strings.NewReader(output))
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		upper := strings.ToUpper(line)
		for _, kw := range doneKeywords {
			if strings.HasPrefix(upper, kw) {
				return "DONE"
			}
		}
		for _, kw := range iterateKeywords {
			if strings.HasPrefix(upper, kw) {
				return "ITERATE"
			}
		}
	}
	return "ITERATE" // default to iterate if ambiguous
}

File: src/loop.ts
import { EventEmitter } from 'events'
import type { AgentRunner, SandboxMode } from './runner.js'
import { renderTemplate, type LoopContext } from './template.js'
import { createSessionLog, appendToLog, logOK, logWarn } from './log.js'
import type { AgentName, StepName } from './config.js'

interface LoopStepConfig {
  agent: AgentName
  model: string
  sandbox: SandboxMode
}

export interface LoopConfig {
  workPrompt: string
  reviewPrompt: string
  gatePrompt: string
  iteratePrompt?: string
  steps: Record<StepName, LoopStepConfig>
  maxIterations: number
  projectRoot: string
  // Seed lastMessage with output from a compound inner node (e.g., repeat before review)
  initialLastMessage?: string
  // Skip the work step on iteration 1 (used when inner node already ran the work)
  skipFirstWork?: boolean
  // Extra template context passed through from executor
  ralphIteration?: number
  maxRalph?: number
  repeatPass?: number
  maxRepeatPasses?: number
}

export interface LoopResult {
  verdict: 'DONE' | 'ITERATE' | 'MAX_ITERATIONS'
  iterations: number
  lastMessage: string
  logFile: string
}

const DONE_KEYWORDS = ['DONE', 'PASS', 'COMPLETE', 'APPROVE', 'ACCEPT']
const ITERATE_KEYWORDS = ['ITERATE', 'REVISE', 'RETRY']

export function parseGateVerdict(output: string): 'DONE' | 'ITERATE' {
  for (const line of output.split('\n')) {
    const upper = line.trim().toUpperCase()
    if (DONE_KEYWORDS.some(kw => upper.includes(kw))) return 'DONE'
    if (ITERATE_KEYWORDS.some(kw => upper.includes(kw))) return 'ITERATE'
  }
  return 'ITERATE'
}

export const loopEvents = new EventEmitter()

export async function agentLoop(
  getRunner: (mode: SandboxMode) => Promise<AgentRunner>,
  config: LoopConfig,
  cookMD: string,
  events: EventEmitter,
): Promise<LoopResult> {
  const logFile = createSessionLog(config.projectRoot)
  events.emit('logFile', logFile)

  let lastMessage = config.initialLastMessage ?? ''

  for (let i = 1; i <= config.maxIterations; i++) {
    // Iteration 1: work → review → gate (or just review → gate if skipFirstWork)
    // Iteration 2+: iterate (or work) → review → gate
    const workStepName: StepName = (i > 1 && config.iteratePrompt) ? 'iterate' : 'work'
    const workPrompt = (i > 1 && config.iteratePrompt) ? config.iteratePrompt : config.workPrompt

    const steps: { name: StepName; prompt: string }[] = (i === 1 && config.skipFirstWork)
      ? [
          { name: 'review', prompt: config.reviewPrompt },
          { name: 'gate', prompt: config.gatePrompt },
        ]
      : [
          { name: workStepName, prompt: workPrompt },
          { name: 'review', prompt: config.reviewPrompt },
          { name: 'gate', prompt: config.gatePrompt },
        ]

    for (const step of steps) {
      const stepConfig = config.steps[step.name]

      events.emit('step', {
        step: step.name,
        iteration: i,
        agent: stepConfig.agent,
        model: stepConfig.model,
      })

      let output: string
      try {
        const ctx: LoopContext = {
          step: step.name,
          prompt: step.prompt,
          lastMessage,
          iteration: i,
          maxIterations: config.maxIterations,
          logFile,
          ralphIteration: config.ralphIteration,
          maxRalph: config.maxRalph,
          repeatPass: config.repeatPass,
          maxRepeatPasses: config.maxRepeatPasses,
        }
        const prompt = renderTemplate(cookMD, ctx)

        events.emit('prompt', prompt)
        const runner = await getRunner(stepConfig.sandbox)
        output = await runner.runAgent(stepConfig.agent, stepConfig.model, prompt, (line) => {
          events.emit('line', line)
        })
      } catch (err) {
        events.emit('error', `${step.name} step failed (iteration ${i}): ${err}`)
        return { verdict: 'ITERATE', iterations: i, lastMessage, logFile }
      }

      lastMessage = output
      try {
        appendToLog(logFile, step.name, i, output)
      } catch (err) {
        console.error(`Warning: failed to write session log: ${err}`)
      }
    }

    const verdict = parseGateVerdict(lastMessage)
    if (verdict === 'DONE') {
      logOK('Gate: DONE — loop complete')
      events.emit('done')
      return { verdict: 'DONE', iterations: i, lastMessage, logFile }
    }
    if (i < config.maxIterations) {
      logWarn(`Gate: ITERATE — continuing to iteration ${i + 1}`)
    }
  }
  logWarn(`Gate: max iterations (${config.maxIterations}) reached — stopping`)
  events.emit('done')
  return { verdict: 'MAX_ITERATIONS', iterations: config.maxIterations, lastMessage, logFile }
}
